File: tools/optimize-images.py
"""Create responsive WebP assets from the site's photographic source files."""

import json
from pathlib import Path

from PIL import Image


PROJECT_ROOT = Path(__file__).resolve().parents[1]
ASSET_DIRECTORY = PROJECT_ROOT / "assets" / "site"
SOURCE_DIRECTORY = PROJECT_ROOT / "images"
SITE_SOURCE_DIRECTORY = SOURCE_DIRECTORY / "source-site"
TARGET_WIDTHS = (640, 960)
MAX_WIDTH = 1200
FACILITY_MAX_WIDTH = 1920
WEBP_QUALITY = 78


def resized(image: Image.Image, target_width: int) -> Image.Image:
    width = min(image.width, target_width)
    if width == image.width:
        return image.copy()
    height = round(image.height * width / image.width)
    return image.resize((width, height), Image.Resampling.LANCZOS)


def convert(source: Path, output_stem: str, max_width: int = MAX_WIDTH) -> None:
    with Image.open(source) as original:
        image = original.convert("RGB")
        main = resized(image, max_width)
        main.save(
            ASSET_DIRECTORY / f"{output_stem}.webp",
            "WEBP",
            quality=WEBP_QUALITY,
            method=6,
        )

        for target_width in TARGET_WIDTHS:
            variant = resized(image, target_width)
            variant.save(
                ASSET_DIRECTORY / f"{output_stem}-{target_width}.webp",
                "WEBP",
                quality=WEBP_QUALITY,
                method=6,
            )


def build_manifest() -> None:
    manifest = {}
    for webp_path in sorted(ASSET_DIRECTORY.glob("*.webp")):
        if webp_path.stem.endswith(("-640", "-960")):
            continue
        with Image.open(webp_path) as image:
            variants = []
            for variant_path in sorted(ASSET_DIRECTORY.glob(f"{webp_path.stem}-*.webp")):
                with Image.open(variant_path) as variant:
                    variants.append({"file": variant_path.name, "width": variant.width})
            variants.append({"file": webp_path.name, "width": image.width})
            manifest[webp_path.name] = {
                "width": image.width,
                "height": image.height,
                "variants": variants,
            }

    (ASSET_DIRECTORY / "image-manifest.json").write_text(
        json.dumps(manifest, indent=2) + "\n",
        encoding="utf-8",
    )


for png_path in sorted(SITE_SOURCE_DIRECTORY.glob("*.png")):
    convert(png_path, png_path.stem)

facility_source = SOURCE_DIRECTORY / "facility-safety-original.jpg"
if facility_source.exists():
    convert(facility_source, "facility-safety", FACILITY_MAX_WIDTH)

build_manifest()

File: tools/optimize-images.mjs
// Create responsive WebP assets from the site's photographic source files.
import { readdir, readFile, stat, writeFile } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import sharp from 'sharp';

const toolDirectory = path.dirname(fileURLToPath(import.meta.url));
const projectRoot = path.resolve(toolDirectory, '..');
const assetDirectory = path.join(projectRoot, 'assets', 'site');
const sourceDirectory = path.join(projectRoot, 'images');
const siteSourceDirectory = path.join(sourceDirectory, 'source-site');

const variantWidths = [640, 960];
const facilityMaxWidth = 1920;
const webpOptions = { quality: 74, effort: 6, smartSubsample: true };
const avifOptions = { quality: 55, effort: 6, chromaSubsampling: '4:2:0' };

// Only these two assets are painted as CSS backgrounds, where an AVIF can be
// offered through image-set() without touching any page markup.
const avifStems = new Set(['campus-learning-lounge', 'facility-safety']);

async function encode(pipeline, targetPath, options, avif) {
  await pipeline.clone().toFormat(avif ? 'avif' : 'webp', avif ? avifOptions : webpOptions).toFile(targetPath);
}

async function convert(source, outputStem, maxWidth) {
  const input = sharp(source, { failOn: 'error' });
  const { width: sourceWidth } = await input.metadata();
  const mainWidth = Math.min(sourceWidth, maxWidth);
  const widths = [...new Set([...variantWidths.filter((width) => width < mainWidth), mainWidth])];

  for (const width of widths) {
    // The sources are soft renders, so sharpen just enough to restore acutance.
    const pipeline = sharp(source).resize({ width, withoutEnlargement: true });
    pipeline.sharpen({ sigma: 0.6, m1: 0.4, m2: 0.25 });

    const suffix = width === mainWidth ? '' : `-${width}`;
    await encode(pipeline, path.join(assetDirectory, `${outputStem}${suffix}.webp`), webpOptions, false);
    if (avifStems.has(outputStem)) {
      await encode(pipeline, path.join(assetDirectory, `${outputStem}${suffix}.avif`), avifOptions, true);
    }
  }
}

async function buildManifest() {
  const files = (await readdir(assetDirectory)).filter((file) => file.endsWith('.webp')).sort();
  const manifest = {};

  for (const file of files) {
    const stem = path.basename(file, '.webp');
    if (/-\d+$/.test(stem)) continue;

    const variants = [];
    for (const variantFile of files.filter((candidate) => new RegExp(`^${stem}-\\d+\\.webp$`).test(candidate))) {
      const { width } = await sharp(path.join(assetDirectory, variantFile)).metadata();
      variants.push({ file: variantFile, width });
    }

    const { width, height } = await sharp(path.join(assetDirectory, file)).metadata();
    variants.sort((left, right) => left.width - right.width);
    variants.push({ file, width });
    manifest[file] = { width, height, variants };
  }

  await writeFile(path.join(assetDirectory, 'image-manifest.json'), `${JSON.stringify(manifest, null, 2)}\n`, 'utf8');
}

for (const file of (await readdir(siteSourceDirectory)).sort()) {
  if (!file.endsWith('.png')) continue;
  await convert(path.join(siteSourceDirectory, file), path.basename(file, '.png'), Number.POSITIVE_INFINITY);
}

const facilitySource = path.join(sourceDirectory, 'facility-safety-original.jpg');
if (await stat(facilitySource).catch(() => null)) {
  await convert(facilitySource, 'facility-safety', facilityMaxWidth);
}

await buildManifest();

const total = (await readdir(assetDirectory))
  .filter((file) => file.endsWith('.webp') || file.endsWith('.avif'));
let bytes = 0;
for (const file of total) bytes += (await stat(path.join(assetDirectory, file))).size;
console.log(`Wrote ${total.length} image files into assets/site (${Math.round(bytes / 1024)} KB).`);
